src/Forms/Label.tsx

```diff
@@ -0,0 +1,92 @@
+import React, {
+  ReactElement,
+  useContext,
+  FunctionComponent,
+} from 'react';
+import styled, { ThemeContext } from 'styled-components';
+import { BaseTheme } from '../Theme/index';
+
+/** An enum that represents the possible values for the label's positioning */
+export enum POSITION {
+  TOP = 'top',
+  LEFT = 'left',
+}
+
+export interface StyledLabelProps {
+  /** Allows you to pass in a label position property from the POSITION enum */
+  labelPosition?: POSITION;
+  /** Specifies the id for the label */
+  htmlFor: string;
+}
+
+export interface StyledLabelTextProps {
+  /** If true, label will be visible */
+  isLabelVisible?: boolean;
+}
+
+export interface LabelProps {
+  /** The id of the field tied to this label */
+  htmlFor: string;
+  /** Specifies the label text */
+  label: string;
+  /** Allows you to pass in a label position property from the POSITION enum */
+  labelPosition?: POSITION;
+  /** If true, label will be visible */
+  isLabelVisible?: boolean;
+}
+
+const StyledLabel = styled.label<StyledLabelProps>`
+  display: grid;
+  grid-template-columns: repeat(3, 1fr);
+  grid-template-rows: ${({ labelPosition }): string => (
+    labelPosition === POSITION.TOP
+      ? 'repeat(3, 1fr)'
+      : 'repeat(2, 1fr)'
+  )};
+  grid-template-areas: ${({ labelPosition }): string => (
+    labelPosition === POSITION.TOP
+      ? `"l l l"
+        "i i i"
+        "e e e"`
+      : `"l i i"
+        ". e e"`
+  )};
```

Comment:
I may have mentioned this, but grid layout can get _super_ complicated quickly. I think this is fundamentally the grid structure we want, but there are a couple of edge cases that we should address:

Basically, This is locking us to either a 3x2 or 3x3 grid, where the input is always going to take up exactly two of the columns, the label will always take up exactly one column, and the rows are the same size. This is OK in a lot of cases, but means that with a short label and wide container, we get this gap between the two:

![2020-03-12_111310](https://user-images.githubusercontent.com/6124793/76546174-6823bf80-6461-11ea-91a1-afa27ecfe830.png)

And if we have a long label in a narrower space, we get this vertical stretching:

![2020-03-12_113918](https://user-images.githubusercontent.com/6124793/76538795-3fe29380-6456-11ea-97ca-470e5931ee13.png)
 
Also, when we have a TextInput that doesn't have a visible label, we get this gap on the left:

![2020-03-12_114233](https://user-images.githubusercontent.com/6124793/76539027-96e86880-6456-11ea-85e3-d747922ed58b.png)

The issues with the wide and narrow columns can actually be fixed with the [`align-items`/`align-self` and `justify-items`/`justify-self` rules that we use with flexbox](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout/Box_Alignment_in_CSS_Grid_Layout), where `justify` affects horizontal alignment, and `align` affects vertical alignment. If we set `align-items: end` on the **grid container**, we can fix the stretching of the input:

![2020-03-12_115700](https://user-images.githubusercontent.com/6124793/76540513-9d77df80-6458-11ea-9ccd-2d4aa74c58e3.png)

but, now we have a huge gap between the input and the error message, since our rows are always going to be the same size. We can change those row definitions to be more flexible with [the `minmax()` function](https://developer.mozilla.org/en-US/docs/Web/CSS/minmax) (functions inside CSS inside javascript :thinking:). I've read those docs a dozen times and I still can't _totally_ explain `minmax`, but essentially it's just setting upper- and lower-bounds on a given size. Here, we can change the row definition line to `grid-template-rows: 1fr minmax(1em, max-content);`. That means we want our first column to be a fixed size, then we want our second column to be no less than `1em` tall, but no greater that the maximum space needed for the content. In practice this means:

With no error message, we keep a blank space

![2020-03-12_120653](https://user-images.githubusercontent.com/6124793/76541505-16c40200-645a-11ea-9812-97a818326f19.png)

When we have a single-line error it takes up that space:

![2020-03-12_120626](https://user-images.githubusercontent.com/6124793/76546213-783b9f00-6461-11ea-8dd3-eea241f174b0.png)

And as the error message grows in length, we switch to max-content:

![2020-03-12_120612](https://user-images.githubusercontent.com/6124793/76546349-b5079600-6461-11ea-9dae-88b9142af1e1.png)

(btw, those red lines are how Firefox displays css-grid elements in the inspector)

Going back to the case where we have a very wide component, we can set `justify-self: end;` on the **LabelText** component to move it over flush with the start of the input:

![2020-03-12_130911](https://user-images.githubusercontent.com/6124793/76546939-b71e2480-6462-11ea-8ce0-4b90c81326b2.png)

> I'd recommend `justify-self`on the element instead of `justify-items` on the container to prevent all of the element from being pushed to the right. You'll also probably want to change that for `POSITION.TOP`.

That's better, but we probably want to provide a little more breathing room between those elements. You could do that with `padding`/`margin`, but css-grid provides [the `gap` property (along with `row-gap` and `column-gap`)](https://developer.mozilla.org/en-US/docs/Web/CSS/gap) to more easily set spacing between the elements of a grid. Setting `gap: 5px 10px` (our theme values for `xsmall` and `small`) will set the `row-gap` and `column-gap` in a single declaration:

![2020-03-12_131427](https://user-images.githubusercontent.com/6124793/76547358-6b1faf80-6463-11ea-9807-3ce3422c7b47.png)

Lastly, we need to handle cases where there is no visible label. This is another tricky one: we don't really want a flexible size here, as we'd want all of the input fields to line up in a form. So the `minmax()` trick from before doesn't work. Probably the best approach would be to add some additional conditions in our `grid-template-areas` declaration. We don't want to use nester ternary operators though, so at this point we probably way to extract that interpolated function into a variable, and create an if-else:

```ts
const generateGrid = (labelPosition, isLabelVisible): string => {
  if (!isLabelVisible) {
    return `"i i i"
            "e e e"`;
  }
  if (labelPosition === POSITION.TOP) {
    return `"l l l"
            "i i i"
            "e e e"`;
  }
  return `"l i i"
          ". e e"`;
};
```

Which you would then need to call from within the styled-components declaration to change the layout.

So that's handling the `POSITION.LEFT` cases, and I suspect there will need to be some additional changes for the `POSITION.TOP` cases. I'm going to leave that you, but let me know if you run into issues with that.

Reply:
Thanks for this super helpful explanation! I believe that all of the styling issues have been fixed/accounted for through 1726e0f.